NeuralNetwork/NeuralNetwork/Matrix.cpp

```diff
@@ -0,0 +1,152 @@
+#include "Matrix.h"
+
+Matrix::Matrix(const int rows, const int columns)
+{
+    assert(columns >= 1 && rows >= 1);
+    _rows = rows;
+    _columns = columns;
+    std::size_t arraySize = columns * rows * sizeof(float);
+    _values = (float*) malloc(arraySize);
+    fill(0.0f);
+}
+
+Matrix::~Matrix()
+{
+    delete [] _values;
+}
+
+float Matrix::get(const int row, const int column) const
+{
+    int index = row * _columns + column;
+    assert(index < (_rows * _columns));
+    return _values[index];
+}
+
+void Matrix::set(const int row, const int column, const float value)
+{
+    assert((row >= 0 && row < _rows) && (column >= 0 && column < _columns));
+    int index = row * _columns + column;
+    _values[index] = value;
+}
+
+int Matrix::getRows() const
+{
+    return _rows;
+}
+
+int Matrix::getColumns() const
+{
+    return _columns;
+}
+
+void Matrix::fill(float value)
+{
+    for (int r = 0; r < _rows; r++)
+    {
+        for (int c = 0; c < _columns; c++)
+        {
+            set(r, c, value);
+        }
+    }
+}
+
+void Matrix::print(std::ostream& stream)
+{
+    print(stream, 3);
+}
+
+void Matrix::print(std::ostream& stream, int decimalPlace)
+{
+    int decimalFactor = pow(10, decimalPlace);
+    assert(decimalFactor > 0);
+    for (int r = 0; r < _rows; r++)
+    {
+        for (int c = 0; c < _columns; c++)
+        {
+            float value = get(r, c);
+            float truncatedValue = floorf(value * decimalFactor) / decimalFactor;
+            stream << "[" << truncatedValue << "] ";
+        }
+        stream << std::endl;
+    }
+}
+
+Matrix* Matrix::clone()
+{
+    Matrix* result = new Matrix(getRows(), getColumns());
+    for (int r = 0; r < getRows(); ++r)
+    {
+        for (int c = 0; c < getColumns(); ++c)
+        {
+            float value = get(r,c);
+            result->set(r,c,value);
+        }
+    }
+    return result;
+}
+
+//STATIC FUNCTIONS
+Matrix* Matrix::multiply(Matrix* const left, Matrix* const right)
+{
+    Matrix* result = nullptr;
+    multiply(left, right, result);
+    return result;
+}
+
+void Matrix::multiply(Matrix* const left, Matrix* const right, Matrix*& output)
+{
+    assert(left != nullptr && right != nullptr);
+    assert(left->getColumns() == right->getRows());
+
+    output = new Matrix(left->getRows(),right->getColumns());
```

Comment:
Would it be better if you do the memory allocation in other place? What if we use only references (`Matrix&`) as parameters?